articles/WideStrings.md

```diff
@@ -0,0 +1,213 @@
+If a message sent from ROS 2 to ROS 1 fails to serialize because a string contains values that are not legal UTF-8 then it won't be published by the bridge.
+
+If a ROS 2 message has a field of type `wstring` then the bridge will attempt to convert it from UTF-16 to UTF-8.
+The resulting UTF-8 encoded string will be published as a `string` type.
+If the conversion fails then the bridge will not publish the message.
+
+## Size of a Wide String
+
+Both UTF-8 and UTF-16 are variable width encodings.
+To minimize the amount of memory used, the `string` and `wstring` types are to be stored in client libraries according to the smallest possible code point.
+This means `string` must be specified as a sequence of bytes, and `wstring` is to be specified as a sequence of words.
+
+Some DDS implementations use 32bit types to store wide strings values.
+This may be due to DDS-XTYPES 1.1 spec saying `wchar` is a 32bit value.
+ROS 2 will instead aim to be compatible with DDS-XTYPES 1.2 and use 16bit storage for wide characters.
+Generated code for ROS 2 messages will automatically handle the conversion when a message is serialized or deserialized.
+
+### Bounded wide strings
+
+Embedded systems accepting strings may use message definitions that restrict the maximum size of a string.
+These are referred to as bounded strings.
+Their purpose is to restrinct the amount of memory used, so the bounds must be specified as units of memory.
+If a `string` field is bounded then the size is given in bytes.
+Similarly the size of a bounded `wstring` is to be specified in words.
+It is the responsibility of whoever populates a bounded `string` or `wstring` to make sure it contains whole code points only.
+Partial code points are indistinuguishable from unknown encodings, so a bounded string whose last code point is incomplete will be published without error.
+
+## Runtime impact of wide string
+
+Dealing with wide strings puts more strain on the software of a system, both in terms of speed and code size.
+UTF-8 and UTF-16 are both variable width encodings, meaning a code point can take 1 (UTF-8 only), 2, or 4 bytes to represent.
+It may take multiple code points to represent a single user perceived character.
+One of the goals of ROS 2 is to support microcontrollers that are contrained by both code size and processor speed.
+Some wide string operations like splitting a string on a user perceived character may not be possible on these devices.
+
+However, whole string equality checking is the same whether using wide strings or not.
+Further splitting a UTF-8 string on an ASCII character is identical to splitting an ASCII character on an ASCII string.
+If a microcontroller must do other types of string manipluation, then a `string` type should still be used.
+Code in the subscriber callback should stop processing a message when it encounters a byte greater than 127 in a `string` field.
```

Comment:
> should

I would rather say "could" and only if it can't handle the encoding.

Reply:
9c4e7b5ab5316d27d62d4e083b035ec27201cd73